seastar/src/config_and_start_seastar.cc

```diff
@@ -0,0 +1,105 @@
+#include <new>
+#include "config_and_start_seastar.hh"
+
+namespace seastar_ffi {
+namespace config_and_start_seastar {
+
+// Copies rust::Slice<const rust::Str> to char**. Returns nullptr if fails.
+// Function free_args should be called on the returned pointer to avoid a memory leak.
+static char** args_as_ptr(rust::Slice<const rust::Str> args) {
+    char** av = new(std::nothrow) char*[args.size() + 1]();
+    if (av == nullptr) {
+        return nullptr;
+    }
+
+    for (size_t i = 0; i < args.size(); i++) {
+        const auto& arg = args[i];
+        av[i] = new(std::nothrow) char[arg.size() + 1];
+        if (av[i] == nullptr) {
+            for (size_t j = 0; j < i; j++) {
+                delete av[j];
+            }
+            delete[] av;
+            return nullptr;
+        }
+        strncpy(av[i], arg.data(), arg.size());
+    }
```

Comment:
I think the strings in `av[i]` might not be null-terminated properly. `strncpy` copies  `n` bytes from the slice, but rust str slices don't contain the null byte.
Adding `()` would zero-initialize the memory:
```c++
av[i] = new(std::nothrow) char[arg.size() + 1]();
```

There might also be utf8 characters that contain a null byte, `strncpy` will stop in the middle of the word then. `memcpy` wouldn't have this issue.

```c
// man strncpy:
// A simple implementation of strncpy() might be:
char* strncpy(char *dest, const char *src, size_t n) {
    size_t i;

    for (i = 0; i < n && src[i] != '\0'; i++)
        dest[i] = src[i];
    for ( ; i < n; i++)
        dest[i] = '\0';

    return dest;
}       
```

Reply:
Good catch. I'll try to RIIR this function - the standard library conversion routines should be aware of those issues.

Reply:
Oh, how convenient, cxx has an example which, more or less, does what we want to do here: https://cxx.rs/binding/rawptr.html